### PR TITLE
test: workqueue: add code description

### PR DESCRIPTION
**Pull request by yingmingx**
1. Modify testcase, add ztest case in test_workq_start_before_submit().
2. Add doxygen comments for details of workqueue.

Signed-off-by: Ying ming <mingx.ying@intel.com>

### CODE REVIEW

File: tests/kernel/workq/work_queue_api/src/main.c [hunk 1]
```diff
@@ -135,19 +164,46 @@ void test_process_work_items_fifo(void)
...
- * - To know that delayed work was executed, we use semaphore.
- * - Right after semaphore was given from handler function, we stop
+ * -# To know that delayed work was executed, we use semaphore.
+ * -# Right after semaphore was given from handler function, we stop
  * measuring actual time.
- * - Then compare results.
+ * -# Then compare results.
+ *
+ * Expected Test Result:
+ * - The measured time results are in a very small range.
```
Comment (LeiW000) [inline tests/kernel/workq/work_queue_api/src/main.c:196]: Here, it's better to specify that range.

**Review by LeiW000 — CHANGES_REQUESTED** [review]

File: tests/kernel/workq/work_queue_api/src/main.c [hunk 2]
```diff
@@ -67,13 +67,42 @@ void common_work_handler(struct k_work *unused)
...
+ * - N/A
+ *
+ * Test Procedure:
+ * -# Creating a work item, then add handler function to that work item.
+ * -# Then process that work item.
+ * -# To check that handler function executed successfully, we use semaphore
+ * sync_sema with initial count 0.
+ * -# Handler function gives semaphore, then we wait for that semaphore
+ * from the test function body.
+ * -# If semaphore was obtained successfully, test passed.
```
Comment (enjiamai) [inline tests/kernel/workq/work_queue_api/src/main.c:92]: Recommend:  "Check" if the semaphore was obtained successfully.

File: tests/kernel/workq/work_queue_api/src/main.c (hunk 1 above)
Comment (enjiamai) [reply tests/kernel/workq/work_queue_api/src/main.c:196]: @yingmingx It seems like the "check point" was not been pointed out in your test procedure, it's a little ambiguous here. 
You could declare it like this:  "Then check if the X-time is within Y and Z".    or  something likes that.

File: tests/kernel/workq/work_queue_api/src/main.c [hunk 3]
```diff
@@ -176,12 +232,41 @@ void test_sched_delayed_work_item(void)
...
+ *
+ * Input Specifications:
+ * - N/A
+ *
+ * Test Procedure:
+ * -#  We can define any number of workqueus using a var of type
+ * struct k_work_q.
+ * -#  Define and initialize maximum possible real number of the
+ * workqueues available according to the stack size.
+ * -#  Test defines and initializes max number of the workqueues
```
Comment (enjiamai) [inline tests/kernel/workq/work_queue_api/src/main.c:255]: Recommend: "Check" would be better here.

File: tests/kernel/workq/work_queue_api/src/main.c [hunk 4]
```diff
@@ -640,13 +725,40 @@ static void ttriggered_work_cancel(const void *data)
...
+ * Prerequisite Conditions:
+ * - N/A
+ *
+ * Input Specifications:
+ * - N/A
+ *
+ * Test Procedure:
+ * -# Call the k_work_q_start() to create a work queue which has it's own
+ * thread that processes the work items in the queue.
+ *
```
Comment (enjiamai) [inline tests/kernel/workq/work_queue_api/src/main.c:748]: no check point here?

File: tests/kernel/workq/work_queue_api/src/main.c [hunk 5]
```diff
@@ -832,20 +944,49 @@ static void work_handler_resubmit(struct k_work *w)
 }
 
 /**
- * @brief Test work submission to queue from handler context, resubmitting
- * a work item during execution of its callback
+ * @brief Test resubmitting work item in callback
  * @details
- * - That test uses sync_sema semaphore with initial count 0.
- * - That test verifies that it is possible during execution of the handler
```
Comment (enjiamai) [inline tests/kernel/workq/work_queue_api/src/main.c:839]: These two sentence looks not like "test procedures", how do you think remove "The test" here?

**Review by enjiamai — CHANGES_REQUESTED** [review]
About the descriptions of procedure, could you please clarify the check point as I comment? Thanks!

File: tests/kernel/workq/work_queue_api/src/main.c (hunk 2 above)
Comment (yingmingx) [reply tests/kernel/workq/work_queue_api/src/main.c:92]: > About the descriptions of procedure, could you please clarify the check point as I comment? Thanks!

Thank you for the reminder. Indeed, I need to write the test steps in this format.

**Comment by yingmingx** [comment]
> About the descriptions of procedure, could you please clarify the check point as I comment? Thanks!

Thank you for the reminder. Indeed, I need to write the test steps in this format.

File: tests/kernel/workq/work_queue_api/src/main.c (hunk 4 above)
Comment (yingmingx) [reply tests/kernel/workq/work_queue_api/src/main.c:753]: I spilt the PR into 2 commits. The first commit use ztest to check this case and add code descriptions in the second commit.

**Comment by LeiW000** [comment]
@yingmingx "buildkite/zephyr" check failed. Please fix it.

**Review by LeiW000 — APPROVED** [review]

**Review by enjiamai — APPROVED** [review]

**Comment by LeiW000** [comment]
@nashif , @andrewboie , @andyross , does anyone help to merge the PR?

**Pull request merged**